### ISSUE TITLE
How to reproduce the results of table 1 (domain generalization experiment)?

### ISSUE THREAD

**Issue by noranta4** [Weixin-Liang/MetaShift#2]
Hello, I am trying to reproduce the results of table 1 in section 4.1, Evaluating domain generalization. 

I have managed to generate the datasets through `domain_generalization_cat_dog.py` (obtaining the directory structure showed in the [readme](https://github.com/Weixin-Liang/MetaShift/#output-files-cat-vs-dog-unmixed-version)) but it is not clear to me how I am supposed to adjust the `main_generalization.py` script to run that domain generalization experiment.

I assume I have to adjust that file from your instructions:

> ### Code for Distribution Shift Experiments
> The python script `experiments/distribution_shift/main_generalization.py` is the entry point for running the distribution shift experiemnts for Section 4.2 (Evaluating Subpopulation Shifts) and Appendix D (Constructing MetaShift from COCO Dataset), and Section 4.1 (Evaluating Domain Generalization). As a running example, the default value for `--data` in `argparse `is `/data/MetaShift/MetaShift-subpopulation-shift` (i.e., for Section 4.2).

In `main_generalization.py` , changing the data argument from `.../data/MetaShift/MetaShift-subpopulation-shift` to `data/MetaShift/Domain-Generalization-Cat-Dog` --i.e. the directory containing the data generated by `domain_generalization_cat_dog.py`-- is not enough, I receive an error about a missing `imageID_to_group.pkl`.  Indeed the data directory is organized differently from the subpopulation-shift experiment (which I successfully reproduced).

Am I missing something?

**Comment by Weixin-Liang** [comment]
Thank you again for your interest in our work. 

We use the implementation of [DomainBed](https://github.com/facebookresearch/DomainBed) for the domain generalization experiment. As you correctly observed, DomainBed requires additional meta-data for the training set. 

More specifically, it requires (1) a meta-data file specifying the 'group' membership information for each training datum (see [here](https://github.com/Weixin-Liang/MetaShift/blob/main/experiments/distribution_shift/main_generalization.py#L71)). This is an example of how to construct such meta-data file ([here](https://github.com/Weixin-Liang/MetaShift/blob/main/dataset/subpopulation_shift_cat_dog_indoor_outdoor.py#L281-L292))

In addition, you also need to specify how do you want to organize the groups into different domains (2) (See [here)](https://github.com/Weixin-Liang/MetaShift/blob/main/experiments/distribution_shift/main_generalization.py#L17-L21). 

So in short, yes you are correct, and you can refer to [DomainBed](https://github.com/facebookresearch/DomainBed) for the input data formats for running your own domain generalization experiments.

**Comment by JunhyunB** [comment]
@Weixin-Liang Can you specify how to build pkl files and the 'DOMAINS_TO_GROUPS' for reproducing Table 1?

**Comment by Weixin-Liang** [comment]
Sure. Take the first cell in Table 1 as an example, with `dog(cabinet + bed)` as dog training data. I. The cat training data is `cat(sofa + bed)` (i.e., `cat(sofa)` unions `cat(bed)`).  

The `DOMAINS_TO_GROUPS` would be: 
```python
DOMAINS_TO_GROUPS = {
            # bed
            0: {'cat':['cat(sofa)'], 'dog':['dog(cabinet)']},
            1: {'cat':['cat(bed)'], 'dog':['dog(bed)']},
}
```
This is one way (though not the only way) to organize the groups into different domains (2) (See [here)](https://github.com/Weixin-Liang/MetaShift/blob/main/experiments/distribution_shift/main_generalization.py#L17-L21). 

The pkl file should contain at least the following keys:
`'cat(sofa)'`, `'cat(bed)'`, `'dog(cabinet)'`, `'dog(bed)'`. 
Again, this is the meta-data file specifying the 'group' membership information for each training datum (see [here](https://github.com/Weixin-Liang/MetaShift/blob/main/experiments/distribution_shift/main_generalization.py#L71)). An example of how to construct such meta-data file can be found [here](https://github.com/Weixin-Liang/MetaShift/blob/main/dataset/subpopulation_shift_cat_dog_indoor_outdoor.py#L281-L292). 

The goal of all these steps is to bridge MetaShift with [DomainBed](https://github.com/facebookresearch/DomainBed). However, since we have already provided the raw image folders, any other ways to format MetaShift for [DomainBed](https://github.com/facebookresearch/DomainBed) would be fine.
